providers/dns/arvancloud/arvancloud.go

```diff
@@ -0,0 +1,182 @@
+func NewDNSProvider() (*DNSProvider, error) {
+	values, err := env.Get(EnvAPIKey)
+	if err != nil {
+		return nil, fmt.Errorf("arvancloud: %w", err)
+	}
+
+	config := NewDefaultConfig()
+	config.APIKey = values[EnvAPIKey]
+
+	return NewDNSProviderConfig(config)
+}
+
+// NewDNSProviderConfig return a DNSProvider instance configured for ArvanCloud.
+func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
+	if config == nil {
+		return nil, errors.New("arvancloud: the configuration of the DNS provider is nil")
+	}
+
+	if config.APIKey == "" {
+		return nil, errors.New("arvancloud: credentials missing")
+	}
+
+	if config.TTL < minTTL {
+		return nil, fmt.Errorf("arvancloud: invalid TTL, TTL (%d) must be greater than %d", config.TTL, minTTL)
+	}
+
+	client := internal.NewClient(config.APIKey)
+
+	if config.HTTPClient != nil {
+		client.HTTPClient = config.HTTPClient
+	}
+
+	return &DNSProvider{config: config, client: client}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS
+// propagation. Adjusting here to cope with spikes in propagation times.
+func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return d.config.PropagationTimeout, d.config.PollingInterval
+}
+
+// Present creates a TXT record to fulfill the dns-01 challenge.
+func (d *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value := dns01.GetRecord(domain, keyAuth)
+
+	authZone, err := getZone(fqdn)
+	if err != nil {
+		return err
+	}
+
+	record := internal.DNSRecord{
+		Type:          "txt",
+		Name:          d.extractRecordName(fqdn, domain),
+		Value:         internal.TXTRecordValue{Text: value},
+		TTL:           d.config.TTL,
+		UpstreamHTTPS: "default",
+		IPFilterMode: &internal.IPFilterMode{
+			Count:     "single",
+			GeoFilter: "none",
+			Order:     "none",
+		},
+	}
+
+	newRecord, err := d.client.CreateRecord(authZone, record)
+	if err != nil {
+		return fmt.Errorf("arvancloud: failed to add TXT record: fqdn=%s: %w", fqdn, err)
+	}
+
+	d.recordIDsMu.Lock()
+	d.recordIDs[token] = newRecord.ID
+	d.recordIDsMu.Unlock()
+
+	return nil
+}
+
+// CleanUp removes the TXT record matching the specified parameters.
+func (d *DNSProvider) CleanUp(domain, token, keyAuth string) error {
+	fqdn, _ := dns01.GetRecord(domain, keyAuth)
+
+	authZone, err := getZone(fqdn)
+	if err != nil {
+		return err
+	}
+
+	// gets the record's unique ID from when we created it
+	d.recordIDsMu.Lock()
+	recordID, ok := d.recordIDs[token]
+	d.recordIDsMu.Unlock()
+	if !ok {
+		return fmt.Errorf("arvancloud: unknown record ID for '%s' '%s'", fqdn, token)
+	}
+
+	if err := d.client.DeleteRecord(authZone, recordID); err != nil {
+		return fmt.Errorf("arvancloud: failed to delate TXT record: id=%s: %w", recordID, err)
+	}
+
+	// deletes record ID from map
+	d.recordIDsMu.Lock()
+	delete(d.recordIDs, token)
+	d.recordIDsMu.Unlock()
+
+	return nil
+}
+
+func getZone(fqdn string) (string, error) {
```

Comment:
just asking, doesn't [effective go](https://golang.org/doc/effective_go.html#Getters) discourages this kind of naming?

Reply:
It's not a real getters (no struct, no method, no attribute related to an attribute)